### PR TITLE
Add support for analyzing using old and new references.

### PR DESCRIPTION
**Pull request by marknfawaz**
## Related issue

Closes: #65
Closes: #66 

## Description
Addresses the referenced issues

## Supplemental testing
Unit testing, integration testing, and manual testing with projects

### CODE REVIEW

**Review by jonlouie — CHANGES_REQUESTED** [review]
Aside from the minor issues in the comments, my main concern is coupling codelyzer to the pre-port data structures. Since not all use cases require pre-port and post-port compilations, what are your thoughts on creating separate classes that handle dual compilation and related APIs?

**Comment by marknfawaz** [comment]
> Aside from the minor issues in the comments, my main concern is coupling codelyzer to the pre-port data structures. Since not all use cases require pre-port and post-port compilations, what are your thoughts on creating separate classes that handle dual compilation and related APIs?

I think we discussed this before and I do share your concern, but decided to go this route to make sure we don't analyze a syntax tree twice and avoid code duplication. Some of these classes will need to be refactored to address this properly. However, just to clarify, we are still supporting the use cases of users analyzing without a legacy compilation.
I'm not sure I understand the separate classes suggestion, can you please clarify? Would that be two projectbuildhandlers? I will create an issue to refactor this and please feel free to provide some ideas to refactor.

File: src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                if (_coreMetaReferences?.Any() == true)
+                {
+                    Compilation = CSharpCompilation.Create(projectName, trees, _coreMetaReferences);
+                }
+            }
+
+            var results = new List<SourceFileBuildResult>();
+
+            _fileInfo.Keys.ToList().ForEach(file => {
+                var sourceFilePath = Path.GetRelativePath(_projectPath, file);
```
[the remark this answers is not in the document]
Comment (marknfawaz) [reply src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs:73]: I'm not sure which line you're referring to here. sourceFilePath is being used in the created object.

File: src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                var fileTree = trees.FirstOrDefault(t => t.FilePath == file);
+
+                var fileResult = new SourceFileBuildResult
+                {
+                    SyntaxTree = fileTree,
+                    PrePortSemanticModel = PrePortCompilation?.GetSemanticModel(fileTree),
+                    SemanticModel = Compilation?.GetSemanticModel(fileTree),
+                    SourceFileFullPath = file,
+                    SourceFilePath = file
+                };
```
[the remark this answers is not in the document]
Comment (marknfawaz) [reply src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs:83]: Good catch. It shouldn't be null, but I added a check.

File: src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonlouie) [reply src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs]: Sorry, GitHub is showing more than what I selected. The variable `sourceFilePath` isn't being used after instantiation. Should it be used in this object creation?
```csharp
var fileResult = new SourceFileBuildResult
{
    SyntaxTree = fileTree,
    PrePortSemanticModel = PrePortCompilation?.GetSemanticModel(fileTree),
    SemanticModel = Compilation?.GetSemanticModel(fileTree),
    SourceFileFullPath = file,
    SourceFilePath = file
};
```

[the remark this answers is not in the document]
Comment (marknfawaz) [reply src/Analysis/Codelyzer.Analysis.Build/FileBuildHandler.cs]: That was a bug, it should've been used in the SourceFilePath field. Just updated.

**Comment by jonlouie** [comment]
> > Aside from the minor issues in the comments, my main concern is coupling codelyzer to the pre-port data structures. Since not all use cases require pre-port and post-port compilations, what are your thoughts on creating separate classes that handle dual compilation and related APIs?
> 
> I think we discussed this before and I do share your concern, but decided to go this route to make sure we don't analyze a syntax tree twice and avoid code duplication. Some of these classes will need to be refactored to address this properly. However, just to clarify, we are still supporting the use cases of users analyzing without a legacy compilation.
> I'm not sure I understand the separate classes suggestion, can you please clarify? Would that be two projectbuildhandlers? I will create an issue to refactor this and please feel free to provide some ideas to refactor.

I wanted to resurface the topic to see if there was additional thought put into this and/or if there was a plan moving forward. I did notice that codelyzer will still work for single compilation apps, but from a maintainability and readability perspective, there would ideally be separate interfaces for cases where users don't need two compilations. For example, based on the changes to `ProjectBuildHandler`, we could extract most of those changes into a child class like:
```csharp
class PortingProjectBuildHandler : ProjectBuildHandler
{
    private Compilation PrePortCompilation;
    private List<string> PrePortMetaReferences;
    private List<string> _metaReferences;
    private string _projectPath;

    public PortingProjectBuildHandler(ILogger logger, AnalyzerConfiguration analyzerConfiguration = null, List<string> metaReferences = null) 
        : base(logger, analyzerConfiguration)
    {
    	_metaReferences = metaReferences;
    }

    private List<PortableExecutableReference> LoadMetadataReferences() { ... }
    private async Task<Compilation> SetPrePortCompilation() { ... }
    public ProjectBuildResult ReferenceOnlyBuild() { ... }
    private Compilation CreateManualCompilation(string projectPath, List<string> references) { ... }
    
    private override async Task SetCompilation()
    {
        PrePortCompilation = await SetPrePortCompilation();  
        base();
    }
}
```
We could use `ProjectBuildHandler` for the usual cases and `PortingProjectBuildHandler` for the porting cases. (I'll add this example to the refactoring issue)

A similar approach can be taken for most other changes, reframing the changes such that we are extending codelyzer rather than modifying it. Doing so would prevent code duplication while isolating business logic based on use case.

**Review by huawenm — APPROVED** [review]

**Review by jonlouie — APPROVED** [review]

**Pull request merged**